### ISSUE TITLE
NVFP4 Training Tracker

### ISSUE THREAD

**Issue by danielvegamyhre** [pytorch/ao#3293]
We want to support NVFP4 training in torchao for both dense and MoE models, following the recipe describe in this [paper](https://arxiv.org/pdf/2509.25149) from NVIDIA. 

<img width="758" height="396" alt="Image" src="https://github.com/user-attachments/assets/63e7725d-5ffe-4d43-99e1-876ea042b934" />

# Support for dense models (linears)

- [ ] Functionality
    - [ ] Random hadamard transform for wgrad gemm inputs (16x16 block granularities)
    - [ ] Rounding modes (https://github.com/pytorch/ao/issues/3264)
        - [ ] Stochastic rounding for **gradients**
            - [ ] Triton wrapping inline PTX e.g., `cvt.rs.satfinite.e2m1x4.f32  d, {a, b, e, f}, rbits; // convert 4 fp32 values to packed 4 e2m1 values with applying .rs rounding`) https://docs.nvidia.com/cuda/parallel-thread-execution/#rounding-modifiers
        - [ ] Round to nearest even (RTNE) for **activations** and **weights** (e.g. `*.rn` modifier)
            - [ ] Triton wrapping inline ptx with `.rn` modifier, same as above
    - [ ] Quantization 
        - [ ] 1x16 quantization for LHS activations (support RTNE and SR as it will be used for grads)
        - [ ] 16x1 quantization for RHS activations (support RTNE, will need SR support if we do `wgrad = input.t() @ grad_out` instead of `wgrad = grad_out.t() @ input` ... may  as well support both rounding modes)
        - [ ] 16x16  quantization for weights, use RNTE
        - [x] 1x16 @ 16x16 scaled gemm
           - [ ] Use for `output = input @ weight.t()`
           - [ ] Use for `dgrad = grad_output @ weight`
        - [x] 1x16 @ 16x1 scaled gemm
           - [ ] Use for `wgrad = grad_output.t() @ input`
    - [ ] Autograd function implementing forward and backward with dynamic quant as described in the paper:
        - [ ] Forward
            - [ ] 1x16 quant with RTNE on 2d input
            - [ ] 16x16 quant with RTNE on 2d weights
            - [ ] 1x16 @ 16x16 scaled mm for output
        - [ ] Backward
            - [ ] dgrad
              - [ ] 1x16 quant with SR on 2d upstream grad
              - [ ] 16x16 quant on 2d weights with RTNE 
              - [ ] per group scale factor conversion to blocked swizzled format (2d grad)
              - [ ] per group scale factor conversion to blocked swizzled format (2d weight)
              - [ ] 1x16 @ 16x16 scaled mm for dgrad
            - [ ] wgrad
              - [ ] RHT on upstream grad (16x16 granularity)
              - [ ] RHT on input activations (16x16 granularity)
              - [ ] 1x16 quant with SR for 2d transposed upstream grad
              - [ ] 16x1 quant with RTNE for 2d input activations
              - [ ] 1x16 @ 16x1 scaled_mm for wgrad  
    - [ ] NVFP4Linear wrapping autograd func (used for module swaps)
    - [ ] DTensor handling for TP support
    - [ ] Custom ops around all custom kernels for `torch.compile` composability
    - [ ] Tests for FSDP, TP
    - [ ] `quantize_` model conversion api peforming module swap of nn.Linear to NVFP4Linear (wraps autograd func)
- [ ] Performance
    - [ ] Random Hadamard transform runs at 80%+ peak achievable memory bandwidth on Blackwell, for all block granularities 
    - [ ] All quantization kernels run at 80%+ of peak achievable memory bandwidth on Blackwell
         - [ ] benchmark scripts for each quantization kernel 
    - [ ] All gemm kernels run at 80%+ of peak achievable TFLOPs/sec on Blackwell 
         - [ ] benchmark scripts for each gemm
- [ ] Integration into torchtitan
    - [ ] Validate loss convergence virtually identical to bf16 for 3k+ steps on full size Llama3 8b/70b 
    - [ ] Validate e2e throughput (TPS) improvement in same training run as above 
- [ ] Documentation
    - [ ] README
    - [ ] torchao docsite  

# Support for MoE layers (grouped GEMMs)
We can extend the low precision MoE training code [here](https://github.com/pytorch/ao/tree/main/torchao/prototype/moe_training) to support NVFP4 by doing the following: 
- [ ] Functionality
    - [ ] Quantization 
      - [ ] 16x16 scaling quantization kernel for 3d expert weights (alternative: update existing kernel to be compatible and performant for 3d inputs)
      - [ ] Triton kernel for per-group conversion to blocked swizzled format where groups are along M (can probably reuse the one we wrote for mxfp8 [here](https://github.com/pytorch/ao/blob/01374eb58a4f9ba52780efbe0ef4e056e36d338c/torchao/prototype/moe_training/kernels/mxfp8/quant.py#L223))
      - [ ] Triton kernel for per-group conversion to blocked swizzled format where groups are along the K/contracting dim (can probably reuse the ones we wrote for mxfp8 [here](https://github.com/pytorch/ao/blob/01374eb58a4f9ba52780efbe0ef4e056e36d338c/torchao/prototype/moe_training/kernels/mxfp8/quant.py#L489)) 
      - [ ] Triton kernel for per-group conversion to blocked swizzled format for 3d expert weights (can probably re-use one we wrote for [mxfp8 here](https://github.com/pytorch/ao/blob/01374eb58a4f9ba52780efbe0ef4e056e36d338c/torchao/prototype/moe_training/kernels/mxfp8/quant.py#L370))
      - [ ] 1x16 @ 16x16 scaled grouped gemm
           - [ ] `output = input @ weight.transpose(-2,-1)`
           - [ ] `dgrad = grad_output @ weight`
      - [ ] 1x16 @ 16x1 scaled grouped gemm
           - [ ] `wgrad = grad_output.transpose(-2,-1) @ input`
    - [ ] Autograd function implementing forward and backward with dynamic quant on inputs (see [mxfp8 example](https://github.com/pytorch/ao/blob/01374eb58a4f9ba52780efbe0ef4e056e36d338c/torchao/prototype/moe_training/scaled_grouped_mm.py#L284))
        - [ ] Forward
            - [ ] 1x16 quant with RTNE on 2d input
            - [ ] 16x16 quant with RTNE on 3d weights
            - [ ] 1x16 @ 16x16 scaled mm for output
        - [ ] Backward
            - [ ] dgrad
              - [ ] 1x16 quant with SR on 2d upstream grad
              - [ ] 16x16 quant on 3d weights with RTNE 
              - [ ] per group scale factor conversion to blocked swizzled format (2d grad)
              - [ ] per group scale factor conversion to blocked swizzled format (3d weight)
              - [ ] 1x16 @ 16x16 scaled mm for dgrad
            - [ ] wgrad
              - [ ] RHT on upstream grad (16x16 granularity)
              - [ ] RHT on input activations (16x16 granularity)
              - [ ] 1x16 quant with SR for 2d transposed upstream grad
              - [ ] 16x1 quant with RTNE for 2d input activations
              - [ ] 1x16 @ 16x1 scaled_mm for wgrad  
    - [ ] Custom ops around all custom kernels for `torch.compile` composability
    - [ ] Tests for FSDP, TP, EP
- [ ] Performance
    - [ ] all quantization kernels run at 80%+ of peak achievable memory bandwidth on Hopper
         - [ ] benchmark scripts for each quantization kernel 
    - [ ] all gemm kernels run at 80%+ of peak achievable TFLOPs/sec on Hopper 
         - [ ] benchmark scripts for each gemm
- [ ] Integration into torchtitan
    - [ ] Validate loss convergence virtually identical to bf16 for 3k+ steps on full size DeepSeekV3 671b 
    - [ ] Validate e2e throughput (TPS) improvement in same training run as above 
 - [ ] Documentation
    - [ ] README
    - [ ] torchao docsite

**Comment by slayton58** [comment]
cc @syed-ahmed for viz

**Comment by namgyu-youn** [comment]
Hi @danielvegamyhre, I would like to work on stochastic rounding support, but I wanted to hear your thoughts before getting started. Could you help with this?

Q1. In my initial understanding, the schema can be decomposed as follows:
```
FP32 Input → Scaling → Rounding → Bit-Packing → Packed NVFP4
```
Is this correct? If so, should I only update the rounding option (RTN/Stochastic) while preserving the common bit-packing layer?

Q2. Based on my previous experience with other subclasses, the changelog would be:
```
to_nvfp4(..., round_mode)

@dataclass
class QuantizeTensorToNVFP4Kwargs(QuantizeTensorKwargs):
    (...)
    round_mode = "stochastic"
```
Is this correct? If not, please let me know where I should add this attribute.

P.S. I haven't looked at the code closely, so I'm not sure how CUDA/Triton/PTX works inside it. It would be greatly appreciated if you could provide some guidance. Thanks.

**Comment by syed-ahmed** [comment]
@namgyu-youn I started a WIP PR last year: https://github.com/pytorch/ao/pull/3384 and am working on finishing it in the next couple weeks.

**Comment by syed-ahmed** [comment]
https://github.com/pytorch/ao/pull/3384 for stochastic rounding is ready for review.

**Comment by rdspring1** [comment]
Created #4040 for random hadamard transform for wgrad gemm inputs (16x16 block granularities)

**Comment by roycho96** [comment]
Hi, can I work on Triton kernels to support for MoE layers?

**Comment by syed-ahmed** [comment]
Update:
- https://github.com/pytorch/torchtitan/pull/3914 demonstrates e2e training on dense models.
- https://github.com/pytorch/ao/pull/4487 improves performance.
- https://github.com/pytorch/ao/pull/4598 adds support for MoEs.

**Comment by krishnakalyan3** [comment]
I'd like to pick up the Documentation items (README + docsite tutorial). Two questions first:

- Is NVFP4TrainingConfig going to be exported and/or moved out of moe_training/? Currently it's [here](https://github.com/pytorch/ao/blob/main/torchao/prototype/moe_training/nvfp4_training/nvfp4_training.py) and only reachable by deep import.
- The [status table](https://docs.pytorch.org/ao/stable/workflows/training.html#nvfp4) still shows performance 🔴 and lists Stochastic rounding/Random Hadamard Transform under planned improvements, though both landed in #4369. Is that stale?.

Happy to start with a PR fixing the status table either way.
